beater/smap_handler_integration_test.go

```diff
@@ -0,0 +1,125 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/apm-server/beater/beatertest"
+	"github.com/elastic/apm-server/beater/headers"
+	"github.com/elastic/apm-server/beater/request"
+	"github.com/elastic/apm-server/tests"
+)
+
+func TestSmapHandler_RequireAuthorizationMiddleware(t *testing.T) {
+	t.Run("Unauthorized", func(t *testing.T) {
+		cfg := cfgEnabledSmap()
+		cfg.SecretToken = "1234"
+		rec := requestToSmapHandler(t, cfg)
+
+		assert.Equal(t, http.StatusUnauthorized, rec.Code)
+		tests.AssertApproveResult(t, assetApprovalPath(t.Name()), rec.Body.Bytes())
+	})
+
+	t.Run("Authorized", func(t *testing.T) {
+		cfg := cfgEnabledSmap()
+		cfg.SecretToken = "1234"
+		h, err := sourcemapHandler(cfg, beatertest.NilReporter)
+		require.NoError(t, err)
+		c, rec := beatertest.ContextWithResponseRecorder(http.MethodPost, "/")
+		c.Request.Header.Set(headers.Authorization, "Bearer 1234")
+		h(c)
+
+		assert.Equal(t, http.StatusBadRequest, rec.Code)
+		tests.AssertApproveResult(t, assetApprovalPath(t.Name()), rec.Body.Bytes())
+	})
+}
+
+func TestSmapHandler_KillSwitchMiddleware(t *testing.T) {
+	t.Run("OffRum", func(t *testing.T) {
+		rec := requestToSmapHandler(t, DefaultConfig(beatertest.MockBeatVersion()))
+
+		assert.Equal(t, http.StatusForbidden, rec.Code)
+		tests.AssertApproveResult(t, assetApprovalPath(t.Name()), rec.Body.Bytes())
+	})
+
+	t.Run("OffSourcemap", func(t *testing.T) {
+		cfg := DefaultConfig(beatertest.MockBeatVersion())
+		rum := true
+		cfg.RumConfig.Enabled = &rum
+		cfg.RumConfig.SourceMapping.Enabled = new(bool)
+		rec := requestToSmapHandler(t, cfg)
+
+		assert.Equal(t, http.StatusForbidden, rec.Code)
+		tests.AssertApproveResult(t, assetApprovalPath(t.Name()), rec.Body.Bytes())
+	})
+
+	t.Run("On", func(t *testing.T) {
+		rec := requestToSmapHandler(t, cfgEnabledSmap())
+
+		assert.Equal(t, http.StatusBadRequest, rec.Code)
+		tests.AssertApproveResult(t, assetApprovalPath(t.Name()), rec.Body.Bytes())
+	})
+}
+
+func TestAssetHandler_LogMiddleware(t *testing.T) {
+	//TODO: How to test?
```

Comment:
maybe just with integration testing is enough?

Reply:
Yes I fear we need to test this with the `system_tests`, tried to cover as much as possible in the go testing, but haven't found a good way to test the `logp.Logger` yet.